### ISSUE TITLE
Support naming tasks after the model name and not alias

### ISSUE THREAD

**Issue by tatiana** [astronomer/astronomer-cosmos#636]
**Context**

Since Cosmos 1.2, we've introduced support to run versioned dbt models (introduced in dbt 1.6), by adopting the alias as the task names, as opposed to the model name. While this works well for versioned tasks, there are use cases when users prefer to stick to the model name as opposed to the alias.

There were recent discussions on the #airflow-dbt slack channel about this:
- by Satyam: https://apache-airflow.slack.com/archives/C059CC42E9W/p1698430202280209?thread_ts=1693472093.058619&cid=C059CC42E9W
```
In one of the scenario we are passing dynamic aliases in config block using the run time parameter named 'division'.
for eg., test_customer_USA or test_customer_CAN where test_customer is my mode file name and USA is division.
I am using manifest to render airflow task and when cosmos render the task group, it uses the alias as the node name (as suggested in this thread) using the default value passed in dbt-project file.
But as per our requirement, the task name should have the model file name instead of alias like it was in last cosmos version
To give you an idea cosmos triggers following command while triggering airflow task:
bin/dbt', 'run', '--vars', "division: CAM\nrun_id: '20231026180212'\n", '--models', 'test_customer_NAM',
here NAM is default parameter passed in dbt_project.yml file and same is used while rendering the airflow task. We need task name as 'test_customer' (same as model file name)
so, need your inputs on how to handle this scenario in cosmos latest version?
```

- by Agnit Chatterjee: https://apache-airflow.slack.com/archives/C059CC42E9W/p1698637532291949
```
Hello everyone,
Hope you are doing well. With the latest cosmos version 1.2.1 we stumbled upon a new problem.
In one of the scenarios we are passing dynamic aliases in config block using the run time parameter named 'division'.
for eg., test_customer_USA or test_customer_CAN where test_customer is the model file name and USA is the division which is passed in as a parameter from the Airflow DAG.
We are using manifest to render airflow task and when cosmos render the task group, it uses the alias as the node name using the default value passed in dbt-project file.
But as per our requirement, the task name should have the model file name instead of alias like it was in last cosmos version
For example,  cosmos triggers following command while triggering airflow task:
bin/dbt', 'run', '--vars', "division: CAM\nrun_id: '20231026180212'\n", '--models', 'test_customer_NAM',
here NAM is default parameter passed in dbt_project.yml file and same is used while rendering the airflow task. We need task name as 'test_customer' (same as model file name)
so, need your inputs on how to handle this scenario in cosmos latest version?
```

**Proposal**

The suggested solution for this problem is introducing a new parameter, `use_alias_as_name`, to `RenderConfig`. By default it will be `True` - so we don't introduce a breaking change. This would allow users to set `use_alias_as_name=False` when they don't want Cosmos to render tasks based on the alias. In those scenarios, model versioning will not work.

**Comment by binhnq94** [comment]
I found the way to made using custom `alias` with/without `var`/`env_var` compatible to versioned models.
Users should config as below (in this example I use `env_var` for easier demo, you can use `var` instead):
```
- name: customers
    description: This table has basic information about a customer, as well as some derived facts based on a customer's orders
    latest_version: 2
    versions:
      - v: 1
        config:
          alias: '{{ "customers_" ~ env_var("DIVISION") ~ "_v1" }}'
      - v: 2
        config:
          alias: '{{ "customers_" ~ env_var("DIVISION") ~ "_v2" }}'
```
And cosmos must to run dbt with node selection as bellow (this is confirmed by dbt engineers [here](https://getdbt.slack.com/archives/C50NEBJGG/p1698827589834889?thread_ts=1698769001.172209&cid=C50NEBJGG) and I tested in the sample dbt project):
```
DIVISION='abc' dbt run --models customers.v1
DIVISION='abc' dbt run --models customers.v2
```
Current version, cosmos will got bug `Not found node` because it run with alias selection as: `--models customers_abc_v1 ` and `--models customers_abc_v2` .
So, how do we update cosmos? How do we get node selection?
I propose to parsing node selection in `unique_id` instead of using `alias` .
```
DIVISION='abc' dbt ls --output json | grep '"resource_type": "model"'

{"name": "customers", "resource_type": "model", "package_name": "jaffle_shop", "original_file_path": "models/customers_v1.sql", "unique_id": "model.jaffle_shop.customers.v1", "alias": "customers_abc_v1", "config": {"enabled": true, "alias": "customers_abc_v1", "schema": null, "database": null, "tags": [], "meta": {}, "group": null, "materialized": "table", "incremental_strategy": null, "persist_docs": {}, "quoting": {}, "column_types": {}, "full_refresh": null, "unique_key": null, "on_schema_change": "ignore", "on_configuration_change": "apply", "grants": {}, "packages": [], "docs": {"show": true, "node_color": null}, "contract": {"enforced": false}, "post-hook": [], "pre-hook": []}, "tags": [], "depends_on": {"macros": [], "nodes": ["model.jaffle_shop.stg_customers", "model.jaffle_shop.stg_orders", "model.jaffle_shop.stg_payments"]}}
{"name": "customers", "resource_type": "model", "package_name": "jaffle_shop", "original_file_path": "models/customers.sql", "unique_id": "model.jaffle_shop.customers.v2", "alias": "customers_abc_v2", "config": {"enabled": true, "alias": "customers_abc_v2", "schema": null, "database": null, "tags": [], "meta": {}, "group": null, "materialized": "table", "incremental_strategy": null, "persist_docs": {}, "quoting": {}, "column_types": {}, "full_refresh": null, "unique_key": null, "on_schema_change": "ignore", "on_configuration_change": "apply", "grants": {}, "packages": [], "docs": {"show": true, "node_color": null}, "contract": {"enforced": false}, "post-hook": [], "pre-hook": []}, "tags": [], "depends_on": {"macros": [], "nodes": ["model.jaffle_shop.stg_customers", "model.jaffle_shop.stg_orders", "model.jaffle_shop.stg_payments"]}}
```
You can see in the logs above, 2 models have the same name and 2 `unique_id` contain `customers.v1` and `customers.v2` .
So node selection should be: `unique_id.split('.', 2)[2]` , reference to [function](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/contracts/graph/node_args.py#L26).
In addition, with this change help cosmos also support versioned models on dbt-core `>=1.5.0` instead `>=1.6.0` as current version.

**Comment by binhnq94** [comment]
I tested it in local, with config:
```
    versions:
      - v: 1
        config:
          alias: '{{ "customers_" ~ var("division") ~ "_v1" }}'
      - v: 2
        config:
          alias: '{{ "customers_" ~ var("division") ~ "_v2" }}'
 ```

```
dbt run --vars "division: 'abc'" -s customers.v1

10:00:10  Running with dbt=1.5.0
10:00:11  Found 6 models, 22 tests, 0 snapshots, 0 analyses, 340 macros, 0 operations, 3 seed files, 0 sources, 0 exposures, 0 metrics, 0 groups
10:00:11  
10:00:13  Concurrency: 10 threads (target='dev')
10:00:13  
10:00:13  1 of 1 START sql table model jaffle_shop_athena.customers_abc_v1 ............... [RUN]
10:00:21  1 of 1 OK created sql table model jaffle_shop_athena.customers_abc_v1 .......... [OK -1 in 8.71s]
10:00:21  
10:00:21  Finished running 1 table model in 0 hours 0 minutes and 10.79 seconds (10.79s).
10:00:21  
10:00:21  Completed successfully
10:00:21  
10:00:21  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1

dbt test --vars "division: 'abc'" -s customers.v1

10:00:33  Running with dbt=1.5.0
10:00:33  Found 6 models, 22 tests, 0 snapshots, 0 analyses, 340 macros, 0 operations, 3 seed files, 0 sources, 0 exposures, 0 metrics, 0 groups
10:00:33  
10:00:39  Concurrency: 10 threads (target='dev')
10:00:39  
10:00:39  1 of 2 START test not_null_customers_v1_customer_id ............................ [RUN]
10:00:39  2 of 2 START test unique_customers_v1_customer_id .............................. [RUN]
10:00:41  2 of 2 PASS unique_customers_v1_customer_id .................................... [PASS in 1.92s]
10:00:41  1 of 2 PASS not_null_customers_v1_customer_id .................................. [PASS in 1.93s]
10:00:41  
10:00:41  Finished running 2 tests in 0 hours 0 minutes and 8.00 seconds (8.00s).
10:00:41  
10:00:41  Completed successfully
10:00:41  
10:00:41  Done. PASS=2 WARN=0 ERROR=0 SKIP=0 TOTAL=2

dbt run --vars "division: 'abc'" -s customers.v2

10:01:44  Running with dbt=1.5.0
10:01:44  Found 6 models, 22 tests, 0 snapshots, 0 analyses, 340 macros, 0 operations, 3 seed files, 0 sources, 0 exposures, 0 metrics, 0 groups
10:01:44  
10:01:48  Concurrency: 10 threads (target='dev')
10:01:48  
10:01:48  1 of 1 START sql table model jaffle_shop_athena.customers_abc_v2 ............... [RUN]
10:01:58  1 of 1 OK created sql table model jaffle_shop_athena.customers_abc_v2 .......... [OK -1 in 10.27s]
10:01:58  
10:01:58  Finished running 1 table model in 0 hours 0 minutes and 14.01 seconds (14.01s).
10:01:58  
10:01:58  Completed successfully
10:01:58  
10:01:58  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1

dbt test --vars "division: 'abc'" -s customers.v2

10:00:50  Running with dbt=1.5.0
10:00:50  Found 6 models, 22 tests, 0 snapshots, 0 analyses, 340 macros, 0 operations, 3 seed files, 0 sources, 0 exposures, 0 metrics, 0 groups
10:00:50  
10:00:53  Concurrency: 10 threads (target='dev')
10:00:53  
10:00:53  1 of 3 START test not_null_customers_v2_customer_id ............................ [RUN]
10:00:53  2 of 3 START test relationships_orders_customer_id__customer_id__ref_customers_  [RUN]
10:00:53  3 of 3 START test unique_customers_v2_customer_id .............................. [RUN]
10:00:56  1 of 3 PASS not_null_customers_v2_customer_id .................................. [PASS in 2.12s]
10:00:56  3 of 3 PASS unique_customers_v2_customer_id .................................... [PASS in 2.13s]
10:00:56  2 of 3 PASS relationships_orders_customer_id__customer_id__ref_customers_ ...... [PASS in 2.13s]
10:00:56  
10:00:56  Finished running 3 tests in 0 hours 0 minutes and 5.09 seconds (5.09s).
10:00:56  
10:00:56  Completed successfully
10:00:56  
10:00:56  Done. PASS=3 WARN=0 ERROR=0 SKIP=0 TOTAL=3
```

**Comment by tatiana** [comment]
I really like this approach, @binhnq94 ! Is there any chance you could contribute it?

**Comment by binhnq94** [comment]
Yes, I will create a PR.

**Comment by jaredx435k2d0** [comment]
I'm not sure if this is related or not, but I'm getting this error:
```Broken DAG: [/usr/local/airflow/dags/dbt_dag.py] Traceback (most recent call last):  File "/usr/local/lib/python3.11/site-packages/airflow/models/baseoperator.py", line 812, in __init__    task_group.add(self)  File "/usr/local/lib/python3.11/site-packages/airflow/utils/task_group.py", line 231, in add    raise DuplicateTaskIdFound(f"{node_type} id '{key}' has already been added to the DAG")airflow.exceptions.DuplicateTaskIdFound: Task id 'INVENTORY_TIER_AND_STORAGE_TYPE_run' has already been added to the DAG```


I think this is because we have some models whose alias is the same as the name of another model. This is because we replicate some of our table models as views in different schemas.

So we may have the following `DB_NAME__SCHEMA_NAME__INVENTORY_TIER_AND_STORAGE_TYPE` model:
```
{{ config(
            materialized='view'
        ,   alias = 'INVENTORY_TIER_AND_STORAGE_TYPE'
        ) 
}}
 

SELECT
    *

FROM 
    {{ ref('INVENTORY_TIER_AND_STORAGE_TYPE') }}  
```

I'm encountering this with Astro 9.5.0 and 1.2.4

**Comment by tatiana** [comment]
@jaredx435k2d0 yes, this seems related.
This fix will be released as part of 1.2.5 this week!
